### PR TITLE
Better mem2reg

### PR DESCRIPTION
**Pull request by ThomasHaas**
This PR (1) adds a new `RemoveDeadNullChecks` pass which is a very simple pass that eliminates null checks and (2) improves `Mem2Reg` to also work on complex allocations (arrays and aggregates).
The `RemoveDeadNullChecks` pass is necessary to enable `Mem2Reg` in some circumstances: `Mem2Reg` only works if the allocated memory is solely used as an address in Loads/Stores, but NULL checks are non-address usages.

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+import static com.dat3m.dartagnan.expression.integers.IntBinaryOp.*;
+import static com.dat3m.dartagnan.expression.integers.IntCmpOp.*;
+
+/*
+    This pass tries to remove unnecessary NULL checks in a very simple manner:
+    it tries to figure out the signedness of expressions and if they are always positive, then they cannot be NULL.
+
+    The signedness is determined as follows:
+      - Registers written to by Allocs as well as MemoryObject expressions are always positive (allocations always succeed!)
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:31]: Should it be "we assume allocation to always succeed"

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    private enum Sign {
+        UNKNOWN,
+        NON_NEG,
+        POS;
+
+        static Sign meet(Sign a, Sign b) {
+            return a.compareTo(b) >= 0 ? b : a;
+        }
+    }
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:51]: This means the order of the enum is important, right? If so, we should document this

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        public Expression visitIntCmpExpression(IntCmpExpr cmp) {
+            if (cmp.getRight() instanceof IntLiteral lit && lit.isZero() && cmp.getLeft().accept(signChecker) == Sign.POS) {
+                // Simplify "expr cop 0" if <expr> is known to be positive.
+                final IntCmpOp op = cmp.getKind();
+                final boolean valueOnNonNull = op == NEQ || op == GT || op == GTE || op == UGT || op == UGTE;
+                return ExpressionFactory.getInstance().makeValue(valueOnNonNull);
+            }
+            return cmp;
+        }
+    }
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:84]: Would it make sense to have a dual pattern for computing expression that are definitely negative?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        public Sign visitIntBinaryExpression(IntBinaryExpr expr) {
+            final Sign leftSign = expr.getLeft().accept(this);
+            final Sign rightSign = expr.getRight().accept(this);
+            if (leftSign == Sign.UNKNOWN || rightSign == Sign.UNKNOWN) {
+                return Sign.UNKNOWN;
+            }
+
+            // --- Both subexpressions are (at least) non-negative ---
+            final IntBinaryOp op = expr.getKind();
+            if (leftSign == Sign.POS && op == ADD) {
```
Comment (hernanponcedeleon) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:151]: Should this be `(leftSign == Sign.POS || rightSign == Sign.POS) && op == ADD`?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java (hunk 1 above)
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:31]: No, allocation events always succeed. This is not directly related to `malloc` which could fail (though we assume it does not).

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java (hunk 3 above)
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:84]: Possible but probably useless.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java (hunk 4 above)
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveDeadNullChecks.java:151]: I guess that would be slightly more accurate

**Pull request merged**